### ISSUE TITLE
Automatic database reconnect

### ISSUE THREAD

**Issue by henning** [go-gorm/gorm#5602]
## Describe the feature

GORM should try to reconnect to a database when connection is lost (e.g. Network error, DB restart)

After the DB restarts, the gorm application usually isnt able to recover and get back to normal operations.
That should be solved.

To reproduce the behaviour that I want to avoid - working with Postgres, I get:

* start your DB
* run you gorm app 
* use your gorm app
* shutdown DB
* try using gorm app
* see it failing with db connection error: "FATAL: terminating connection due to administrator command (SQLSTATE 57P01)"
* restart DB
* Try using App
* See it still getting the same error

Expecation would be that the application reconnects to the DB and continue to work without a restart.

## Motivation

GORM applications without DB connection do not work very well.

## Related Issues

Multiple issues have been created with that, but all have been closed, but without mentioning a reason or a workaround:

https://github.com/go-gorm/gorm/issues?q=reconnect

There are also some chat messages about it:  https://gitter.im/jinzhu/gorm?at=5a59df94ba39a53f1a0e36f2 - but it's not clear to me why gorm doesn't come out of the box with that reconnect behaviour - are there important downsides to configure it by default to reconnect automatically? 

I'd be happy to try find a solution for that thing or add some documentation with a workaround if I get some guidance for more experienced gorm developers about how to do it best... maybe there is a reason why it has not been implemented before, or maybe I am overseeing something that makes it not a useful feature.

**Comment by sincerefly** [comment]
I need this feature

**Comment by borzovplus** [comment]
any updates?

**Comment by flaxinger** [comment]
I just started googling this topic and landed on this stale issue.
Any hint to prebuilt solutions would be of much help (if there are any)

**Comment by nk915** [comment]
I need this feature too.

**Comment by 2015WUJI01** [comment]
Is there any solution now?

**Comment by Stogas** [comment]
I am unable to replicate this in a Gin API application, i.e. it recovers automatically with a subsequent request.

has this feature been implemented?

**Comment by lyn1505** [comment]
gorm automatically reconnects for me. I'm using gorm version v1.25.12.

**Comment by dsychin** [comment]
I am running into the same issue on v1.25.12 with mysql v1.8.1.
Code is set up to connect using an [existing sql connection.](https://gorm.io/docs/connecting_to_the_database.html#Existing-database-connection)
Previously I had gorm v1.24.2 with mysql 1.4.4 and it would reconnect automatically but now once it fails it will never reconnect.

```
{"time":"2024-11-05T10:39:56.01449173+09:00","level":"ERROR","prefix":"-","file":"gormlogger.go","line":"44","message":"failed to initialize database, got error %v[dial tcp [::1]:3307: connect: connection refused]"}
{"time":"2024-11-05T10:40:04.352576548+09:00","level":"ERROR","prefix":"-","file":"gormlogger.go","line":"44","message":"failed to initialize database, got error %v[sql: database is closed]"}
{"time":"2024-11-05T10:40:12.343878406+09:00","level":"ERROR","prefix":"-","file":"gormlogger.go","line":"44","message":"failed to initialize database, got error %v[sql: database is closed]"}
```

**Comment by jinzhu** [comment]
> gorm automatically reconnects for me. I'm using gorm version v1.25.12. It would be helpful if @jinzhu can confirm this

> Reference in

Yes, Go manages the connection pool and will handle reconnections after a failure.

**Comment by nk915** [comment]
@jinzhu Thank you.

**Comment by dsychin** [comment]
@jinzhu I believe the reconnection feature is broken when using [an existing database connection](https://gorm.io/docs/connecting_to_the_database.html#Existing-database-connection).

The issue seems to be caused by [this change](https://github.com/go-gorm/gorm/pull/6249/files) released with v1.25.

When using [gorm.Open with a dsn](https://gorm.io/docs/connecting_to_the_database.html#MySQL), this is not an issue because [a new connection would be created](https://github.com/go-gorm/mysql/blob/master/mysql.go#L119-L126), however when using an existing database connection, the closed connection is reused which would fail.

I don't think this is the intended behaviour?

**Comment by jinzhu** [comment]
> @jinzhu I believe the reconnection feature is broken when using [an existing database connection](https://gorm.io/docs/connecting_to_the_database.html#Existing-database-connection).
> 
> The issue seems to be caused by [this change](https://github.com/go-gorm/gorm/pull/6249/files) released with v1.25.
> 
> When using [gorm.Open with a dsn](https://gorm.io/docs/connecting_to_the_database.html#MySQL), this is not an issue because [a new connection would be created](https://github.com/go-gorm/mysql/blob/master/mysql.go#L119-L126), however when using an existing database connection, the closed connection is reused which would fail.
> 
> I don't think this is the intended behaviour?

This doesn’t make any difference, GORM handles it the same way internally.
